### PR TITLE
Fixes #374 : Use strict instead of lenient date parsing by default

### LINKED ISSUE #374: Use strict instead of lenient date parsing mode by default
While parsing String representing the date, [setLenient(false)](http://docs.oracle.com/javase/7/docs/api/java/text/DateFormat.html#setLenient(boolean)) mode should be used.

It solves some date parsing anomalies, such as allowing January 32nd to be interpreted as February first or even January -1st as December 31st. More examples can be found [here](http://stackoverflow.com/questions/7606387/what-is-the-use-of-lenient).

The commit message should say that it's a braking change so that when I do the release notes I won't forget it.

We need a way to set the parsing leniency so that users can switch to the old behaviour. The (new) default behaviour being not to be lenient.

##### Design question
* Should this be static and global AssertJ parameter? Then we need to add a method `setStrictDateParsing(boolean)` in Assertions delegating the real work to the same method in `AbstractDateAssert`.
* Otherwise we may add e.g. `usingLenientDateParsing()` to `AbstractDateAssert` to parse date leniently just for one particular assertion.

### PR DESCRIPTION
**Pull request by mkordas**
This change breaks backward compatibility.

### CODE REVIEW

File: src/main/java/org/assertj/core/api/AbstractDateAssert.java [hunk 1]
```diff
@@ -2345,6 +2345,45 @@ public S withDateFormat(String userCustomDateFormatPattern) {
...
+   * assertThat(date).hasSameTimeAs("2001-02-04"); // different date
+   * assertThat(dateTime).hasSameTimeAs("2001-02-03 04:05:06"); // leniency does not help here
+   * </code></pre>
+   *
+   * To revert to default strict date parsing, call {@code useLenientDateParsing(false)}.
+   *
+   * @param value whether lenient parsing mode should be enabled or not
+   * @author Michal Kordas
+   */
+  public static void useLenientDateParsing(boolean value) {
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/api/AbstractDateAssert.java]: I prefer to rename `useLenientDateParsing` to `setLenientDateParsing`.
When I read `useLenientDateParsing` the first thing that comes to my mind is that the parsing will be lenient.

File: src/main/java/org/assertj/core/api/Assertions.java [hunk 2]
```diff
@@ -1193,6 +1193,43 @@ public static String contentOf(File file) {
...
+   * assertThat(date).hasSameTimeAs("2001-02-04"); // different date
+   * assertThat(dateTime).hasSameTimeAs("2001-02-03 04:05:06"); // leniency does not help here
+   * </code></pre>
+   *
+   * To revert to default strict date parsing, call {@code useLenientDateParsing(false)}.
+   *
+   * @param value whether lenient parsing mode should be enabled or not
+   * @author Michal Kordas
+   */
+  public static void useLenientDateParsing(boolean value) {
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/api/Assertions.java]: Rename to `setLenientDateParsing`.

File: src/main/java/org/assertj/core/api/AbstractDateAssert.java (hunk 1 above)
Comment (mkordas) [inline src/main/java/org/assertj/core/api/AbstractDateAssert.java]: Agreed, done

**Comment by joel-costigliola** [comment]
Integrated, thanks.

**Pull request closed without merging**